producer/src/contract/contract.go

```diff
@@ -38,65 +39,114 @@ func InsertYield(y Yield, database string, blockHeight uint32, contractHash []by
 		fmt.Println("Failed to insert")
 		return err2
 	}
-
-	dbConn.Close()
 	return nil
 }
 
-/* Serialize ... serialies the yield */
+/*Serialize serializes the yield */
 func (y *Yield) Serialize() []byte {
 	s := make([]byte, 40) //32 bytes for hash and 8 bytes for value
 	copy(s[0:32], y.Recipient)
 	binary.LittleEndian.PutUint64(s[32:40], y.Value)
 	return s
 }
 
-/* DeserializeYield ... deserializes the yield */
+/*DeserializeYield deserializes the yield */
 func DeserializeYield(b []byte) Yield {
 	recipient := make([]byte, 32)
 	copy(recipient, b[0:32])
 	value := binary.LittleEndian.Uint64(b[32:40])
 	return Yield{Recipient: recipient, Value: value}
 }
 
-/*
-Contains the contract hash of the claimed yield,
-the block index containing the contract of the claimed yield,
-the index of the claimed yield in the contract,
-and the public key of the claimant
-*/
-type Claim struct{}
-
-/*
-Should scan the Unclaimed Yield Pool for a yield
-
-Prioritize yields that are closest to the value parameter,
-ie. MIN(abs(value - yieldValue0), abs(value - yieldValue1), ... abs(value - yieldValueN))
-
-Case 1:
-If the claimed yield is less than or equal to the value, return
-the claim and a nil for the error
-
-Case 2:
-If the claimed yield exceeds the value, return
-the claim as usual and a custom error struct with
-the difference as a uint64 field called "Change"
-
-Case 3:
-If there are no yields left in the Pool,
-return an empty Claim struct and a custom error struct
-that simply states there are insufficient funds
-*/
+/*Claim contains information about which yield it is claiming*/
+type Claim struct {
+	PreviousContractHash []byte
+	BlockIndex           uint64
+	YieldIndex           uint8
+	PublicKey            ecdsa.PublicKey
+}
+
+/*MakeClaim will query the database to find the best yield to claim */
 func MakeClaim(database string, claimant ecdsa.PublicKey, value uint64) (Claim, error) {
-	return Claim{}, errors.New("Incomplete function")
+	claimantHexHash := hex.EncodeToString(block.HashSHA256(keys.EncodePublicKey(&claimant)))
+	//open datase first
+	dbConn, err := sql.Open("sqlite3", database)
+	if err != nil {
+		log.Printf("Error! unable to open databse: %v\n", err)
+		return Claim{}, err
+	}
+	defer dbConn.Close()
+
+	var dbHeight uint64
+	var dbContract string
+	var dbIndex uint8
+	var dbHolder string
+	var dbValue uint64
+
+	//first query the db for any yeilds equal or above
+	sqlQuery := `SELECT * FROM uy WHERE value >= $1 AND holder = $2 ORDER BY value ASC LIMIT 1;`
+	rows, err := dbConn.Query(sqlQuery, value, claimantHexHash)
```

Comment:
We should  make one query to grab all yields, then find the mininimum absolute difference between the claimant desired value and the value of all the yields.

Reply:
Why not find a yield that can fully pay for the claim?

Reply:
And then if that fails, find yield that can pay the most.

Reply:
Because the smallest possible absolute minimum difference is zero. The yield selection part should be achievable with a single query. Also the chances of getting exact yield values are less and less likely the more you make contracts. Doing the absolute minimum difference means that if there's a yield that is exact, we will grab it,  otherwise we grab whatever is nearest to it. If your bill is 15, and you have a 10, 10, and a 100, you'd probably use claim the 10 then the other 10 to make your change small. If you had a 15 available the absolute minimum difference would pick that instead with one query.

Reply:
But the first query tries to get yields that are equal to or greater than the desired value. If there are no yields that can fully fund the claim, only then is the second query executed.